subprojects/base-services/src/main/java/org/gradle/internal/work/DefaultWorkerLeaseService.java

```diff
@@ -101,6 +101,17 @@ public DefaultWorkerLease getWorkerLease() {
     }
 
     @Override
+    public void withSharedLease(WorkerLease sharedLease, Runnable action) {
```

Comment:
❌ Let's put some unit testing for this in `DefaultWorkerLeaseServiceWorkerLeaseTest`.  Minimally, I would test that a child lease created in the action borrows from the parent and that a child lease in the action blocks when there are no more leases available (ie the parent has already been borrowed).